# Python for Life Scientists
This course is a Python introduction for all life scientists. Many examples have a focus on genomics and sequence analysis. Please feel free to re-use the material based on terms of a CC-BY license.

Please find additional data sets for exercises in the data_examples subfolder.

Please find additional scripts here (original Python course developed by Boas Pucker at CeBiTec / Bielefeld University):
https://github.com/bpucker/APPLS



# References

Pucker B., Schilbert H. M., and Schumacher S. F. (2019). Integrating Molecular Biology and Bioinformatics Education. Journal of Integrative Bioinformatics, vol. 16, no. 3, 2019, pp. 20190005. doi: [10.1515/jib-2019-0005](https://doi.org/10.1515/jib-2019-0005)

Pucker, B., Holtgräwe, D., Rosleff Sörensen, T., Stracke, R., Viehöver, P., and Weisshaar, B. (2016). A de novo Genome Sequence Assembly of the Arabidopsis thaliana Accession Niederzenz-1 Displays Presence/Absence Variation and Strong Synteny. PloS-ONE 11:e0164321. doi: [10.1371/journal.pone.0164321](https://doi.org/10.1371/journal.pone.0164321).

Pucker, B., Holtgräwe, D., & Weisshaar, B. (2017). Consideration of non-canonical splice sites improves gene prediction on the Arabidopsis thaliana Niederzenz-1 genome sequence. BMC Research Notes, 10, 667. doi: [10.1186/s13104-017-2985-y](https://doi.org/10.1186/s13104-017-2985-y).

Haak, M., Vinke, S., Keller, W., Droste, J., Rückert, C., Kalinowski, J., & Pucker, B. (2018). High Quality de novo Transcriptome Assembly of Croton tiglium. Frontiers in Molecular Biosciences, 5. doi: [10.3389/fmolb.2018.00062](https://doi.org/10.3389/fmolb.2018.00062).
