# Data and examples for 'Python for Life Scientists'

This collection of data and examples is part of the 'Python for Life Scientists' course run by @bpucker / @PuckerLab.

AtCol0_Exons.fasta.7z (needs to be extracted before use). Source: [TAIR10](https://www.arabidopsis.org/download/)/Araport11

barplots.py

boxplot.py

construct_heatmap.py

Ct.cds.fasta.7z (needs to be extracted before use). Source: Haak et al., 2018: [10.3389/fmolb.2018.00062](https://doi.org/10.3389/fmolb.2018.00062)

expression_data.ods

gene_structure_plot.py

genetic_code.csv

heatmap.html

hist_indel_lengths.py

load_blast_results.py

results.txt

scatter_plot.py

UNKNOWN_DATA.ods

## References

Pucker, B., Holtgräwe, D., Rosleff Sörensen, T., Stracke, R., Viehöver, P., and Weisshaar, B. (2016). A de novo Genome Sequence Assembly of the Arabidopsis thaliana Accession Niederzenz-1 Displays Presence/Absence Variation and Strong Synteny. PloS-ONE 11:e0164321. doi: [10.1371/journal.pone.0164321](https://doi.org/10.1371/journal.pone.0164321).

Pucker, B., Holtgräwe, D., & Weisshaar, B. (2017). Consideration of non-canonical splice sites improves gene prediction on the Arabidopsis thaliana Niederzenz-1 genome sequence. BMC Research Notes, 10, 667. doi: [10.1186/s13104-017-2985-y](https://doi.org/10.1186/s13104-017-2985-y).

Haak, M., Vinke, S., Keller, W., Droste, J., Rückert, C., Kalinowski, J., & Pucker, B. (2018). High Quality de novo Transcriptome Assembly of Croton tiglium. Frontiers in Molecular Biosciences, 5. doi: [10.3389/fmolb.2018.00062](https://doi.org/10.3389/fmolb.2018.00062).
